Add SecureBoot support for arbitrary operating systems to "Grub2 UEFI" PXE loaders (#2145)
theforeman/foreman-documentation: guides/common/modules/proc_configuring-smart-proxy-to-provision-secure-boot-enabled-hosts.adoc

----
+
You can download the `{grub_efi_package_name}` package from {grub_efi_download_url}.
You can download the `{shim_efi_package_name}` package from {shim_efi_download_url}.
. Extract the shim and GRUB2 binaries:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# {extract_grub}
# {extract_shim}
----
. Make the shim and GRUB2 binaries available for host provisioning:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# cp {grub_efi_tmp_binary_path} $BOOTLOADER_PATH/grubx64.efi
# cp {shim_efi_tmp_binary_path} $BOOTLOADER_PATH/shimx64.efi
# ln -sr $BOOTLOADER_PATH/grubx64.efi $BOOTLOADER_PATH/boot.efi
# ln -sr $BOOTLOADER_PATH/shimx64.efi $BOOTLOADER_PATH/boot-sb.efi
# chmod 644 $BOOTLOADER_PATH/grubx64.efi
# chmod 644 $BOOTLOADER_PATH/shimx64.efi
----
ifeval::["{client-pkg-ext}" == "deb"]
. Link the `grub.cfg` file from the TFTP servers `grub2` folder to the legacy `grub` folder:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# ln --relative --symbolic /var/lib/tftpboot/grub2/grub.cfg /var/lib/tftpboot/grub/grub.cfg
----
endif::[]

.Verification
* Verify the contents of your boot loader directory:
+

> [options="nowrap" subs="+quotes,verbatim,attributes"]

Reviewer: ```suggestion
[source, none, options="nowrap" subs="+quotes,verbatim,attributes"]
```

Refs 666b0821422caee657cb01c6bd1e63edb29dd20d

[id="configuring-{smart-proxy-context}-to-provision-{client-os-context}-on-Secure-Boot-enabled-hosts"]

> = Configuring {SmartProxy} to provision {client-os} on Secure Boot enabled hosts

Reviewer 2: ```suggestion
= Configuring {SmartProxy} to provision {client-os} on Secure Booted hosts
```
Would this be acceptable? It would make the title shorter.

Reviewer 3: We should leave this as is. Hosts with enabled "Secure Boot" are "securely booted", that's right. But "Secure Boot" is the proper name for the mechanism and the protocol used to achieve this and the term is part of the [UEFI specification](https://uefi.org/sites/default/files/resources/UEFI_Spec_2_10_Aug29.pdf) (see page 1411).

But i have to admit the title is pretty long :see_no_evil:

Reviewer 2: That's okay. I don't have any other ideas how to make it shorter :laughing:

Author: "[...] on Secure Booted hosts" doesn't make any sense here. Please keep "[...] Secure Boot enabled hosts" as suggested by @goarsna.

Secure Boot follows a chain of trust from the start of the host to the loading of Linux kernel modules.
The first shim that is loaded determines which distribution can be booted or loaded by using a `kexec` system call until the next reboot.

To provision {client-os} on Secure Boot enabled hosts with the *Grub2 UEFI SecureBoot* and *Grub2 UEFI HTTPS SecureBoot* PXE loaders, you have to provide signed shim and GRUB2 binaries provided by the vendor of your operating system.

[IMPORTANT]
====
ifdef::satellite[]
You have to perform the following configuration steps on each TFTP {SmartProxy} for a subnet to provision Secure Boot enabled hosts on that subnet.
endif::[]
ifndef::satellite[]

> You have to perform the following configuration steps on each TFTP proxy for a subnet to provision Secure Boot enabled hosts on that subnet.

Reviewer 2: ```suggestion
You have to perform the following configuration steps on each TFTP {SmartProxy} for a subnet to provision Secure Boot enabled hosts on that subnet.
```
Is there a reason not to call it SmartProxy in all builds?

Reviewer: out of scope: in orcharhino, we have recently changes/fixed the branding in the WebUI. Changing this should not be part of this PR. Instead, I will eventually create a follow-up PR to remove some "ifdef::orcharhino -> use X proxy" in favor of "use X {SmartProxy}".

Please align with current docs.

endif::[]
====

The following example works for {client-os} on x86_64 architecture.

.Prerequisites

> ifeval::["{client-os}" == "Debian"]
> * Ensure that `ar` and `xz` are installed on your {SmartProxy}.
> endif::[]
> ifeval::["{client-os}" == "Ubuntu"]
> * Ensure that `ar`, `xz`, and `zstd` are installed on your {SmartProxy}.
> endif::[]
> ifeval::["{client-pkg-ext}" == "rpm"]
> * Ensure that `cpio` is installed on your {SmartProxy}.
> endif::[]

Reviewer 2: ```suggestion
ifeval::["{client-os}" == "Debian"]
* Ensure that the `ar` and `xz` packages are installed on your {SmartProxy}.
endif::[]
ifeval::["{client-os}" == "Ubuntu"]
* Ensure that the `ar`, `xz`, and `zstd` packages are installed on your {SmartProxy}.
endif::[]
ifeval::["{client-pkg-ext}" == "rpm"]
* Ensure that the `cpio` package is installed on your {SmartProxy}.
endif::[]
```
If they are package names, I'd specify that we're talking about packages.

----
+
You can download the `{grub_efi_package_name}` package from {grub_efi_download_url}.
You can download the `{shim_efi_package_name}` package from {shim_efi_download_url}.
. Extract the shim and GRUB2 binaries:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# {extract_grub}
# {extract_shim}
----
. Make the shim and GRUB2 binaries available for host provisioning:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# cp {grub_efi_tmp_binary_path} $BOOTLOADER_PATH/grubx64.efi
# cp {shim_efi_tmp_binary_path} $BOOTLOADER_PATH/shimx64.efi
# ln -sr $BOOTLOADER_PATH/grubx64.efi $BOOTLOADER_PATH/boot.efi
# ln -sr $BOOTLOADER_PATH/shimx64.efi $BOOTLOADER_PATH/boot-sb.efi
# chmod 644 $BOOTLOADER_PATH/grubx64.efi
# chmod 644 $BOOTLOADER_PATH/shimx64.efi
----
ifeval::["{client-pkg-ext}" == "deb"]
. Link the `grub.cfg` file from the TFTP servers `grub2` folder to the legacy `grub` folder:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# ln --relative --symbolic /var/lib/tftpboot/grub2/grub.cfg /var/lib/tftpboot/grub/grub.cfg
----
endif::[]

.Verification

> * Verify the contents of your bootloader directory:

Reviewer 3: @apinnick: `vale` complains about the spelling of "bootloader". Searching in foreman-documentation I found approximately 300 appearances of "bootloader" and 300 of "boot loader". Which way should we go?

Reviewer 4: Hi @goarsna, you can go with what Vale suggests, so "boot loader". For the record, it looks like this particular rule comes from the [RH Supplementary Style guide](https://redhat-documentation.github.io/supplementary-style-guide/#boot-loader).

Reviewer: This will be fixed with https://github.com/theforeman/foreman-documentation/pull/3274

The first shim that is loaded determines which distribution can be booted or loaded by using a `kexec` system call until the next reboot.

To provision {client-os} on Secure Boot enabled hosts with the *Grub2 UEFI SecureBoot* and *Grub2 UEFI HTTPS SecureBoot* PXE loaders, you have to provide signed shim and GRUB2 binaries provided by the vendor of your operating system.

[IMPORTANT]
====
ifdef::satellite[]
You have to perform the following configuration steps on each TFTP {SmartProxy} for a subnet to provision Secure Boot enabled hosts on that subnet.
endif::[]
ifndef::satellite[]
You have to perform the following configuration steps on each TFTP proxy for a subnet to provision Secure Boot enabled hosts on that subnet.
endif::[]
====

The following example works for {client-os} on x86_64 architecture.

.Prerequisites
ifeval::["{client-os}" == "Debian"]
* Ensure that `ar` and `xz` are installed on your {SmartProxy}.
endif::[]
ifeval::["{client-os}" == "Ubuntu"]
* Ensure that `ar`, `xz`, and `zstd` are installed on your {SmartProxy}.
endif::[]
ifeval::["{client-pkg-ext}" == "rpm"]
* Ensure that `cpio` is installed on your {SmartProxy}.

> endif::[]

Reviewer 5: Why don't we make these packaging dependencies if they're required?

> .Procedure

Reviewer 5: If this is all common, why don't we ship a helper script that takes several inputs and saves it to the correct places? That makes both the procedure and testing easier. We can also ensure all commands we use are properly ensured via packaging dependencies.

There's precedent for this: https://github.com/theforeman/smart-proxy/blob/develop/sbin/foreman-prepare-realm is installed ([here](https://github.com/theforeman/foreman-packaging/blob/067c3b95706b2ebf112aa4cf096b39caab6778b4/packages/foreman/foreman-proxy/foreman-proxy.spec#L135) for RPM). Then the documentation refers to this:
https://github.com/theforeman/foreman-documentation/blob/3259af88086c536d683e3d2846172ca5051d5b4d/guides/common/modules/con_configuring-project-to-manage-the-lifecycle-of-a-host-registered-to-a-freeipa-realm.adoc#L29-L33

----
# install -o foreman-proxy -g foreman-proxy -d $BOOTLOADER_PATH
----
. Download the shim and GRUB2 packages for the operating system of your host:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# wget -O /tmp/{grub_efi_downloaded_package_name} _https://{server-example-com}/{grub_efi_downloaded_package_name}_
# wget -O /tmp/{shim_efi_downloaded_package_name} _https://{server-example-com}/{shim_efi_downloaded_package_name}_
----
+
You can download the `{grub_efi_package_name}` package from {grub_efi_download_url}.
You can download the `{shim_efi_package_name}` package from {shim_efi_download_url}.
. Extract the shim and GRUB2 binaries:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# {extract_grub}
# {extract_shim}
----
. Make the shim and GRUB2 binaries available for host provisioning:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# cp {grub_efi_tmp_binary_path} $BOOTLOADER_PATH/grubx64.efi
# cp {shim_efi_tmp_binary_path} $BOOTLOADER_PATH/shimx64.efi
# ln -sr $BOOTLOADER_PATH/grubx64.efi $BOOTLOADER_PATH/boot.efi
# ln -sr $BOOTLOADER_PATH/shimx64.efi $BOOTLOADER_PATH/boot-sb.efi
# chmod 644 $BOOTLOADER_PATH/grubx64.efi

> # chmod 644 $BOOTLOADER_PATH/shimx64.efi

Reviewer 3: I would suggest to leave this as it is. For the bootloader universe to work it is necessary that the bootloader files can be read by the `foreman-proxy` user. This is achieved by setting the permissions to `644`. The files should never be altered by the `foreman-proxy` user.

> # chmod 644 $BOOTLOADER_PATH/shimx64.efi

Reviewer 6: Let me re-test that to ensure it's working properly.

> # chmod 644 $BOOTLOADER_PATH/shimx64.efi

Reviewer: Any updates on the permissions @nofaralfasi ?

> # chmod 644 $BOOTLOADER_PATH/shimx64.efi

Reviewer 6: It is not needed.

----
ifeval::["{client-pkg-ext}" == "deb"]
. Link the `grub.cfg` file from the TFTP servers `grub2` folder to the legacy `grub` folder:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----

> # ln -rs /var/lib/tftpboot/grub2/grub.cfg /var/lib/tftpboot/grub/grub.cfg

Reviewer 5: TIL: `-r` to make a link relative.

I always prefer long options to make it self-documenting

```suggestion
# ln --relative --symbolic /var/lib/tftpboot/grub2/grub.cfg /var/lib/tftpboot/grub/grub.cfg
```

----
endif::[]

.Verification
* Verify the contents of your bootloader directory:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# tree _My_Bootloader_Directory_
_My_Bootloader_Directory_
└── pxegrub2

>     ├── _{secureboot-os-name}_

Reviewer 5: This looks copy-pasted from a more complex tree
```suggestion
    └── _{secureboot-os-name}_
```
